### ISSUE TITLE
tabbing at the keyboard, focus outlines, and button order

### ISSUE THREAD

**Issue by albell** [videojs/video.js#1500]
If you go to the homepage and play the vid, and then pause it, and start tabbing around, the focus moves in markup order, like this:

- play button
- progress control

…so far so good. But then it leaps to the far right and starts incrementing to the left:

- fullscreen
- volume
- mute
- captions

This feels like a friction point to me. The first time I did this, I had no idea where focus had leapt to. I had to do it over and over again, watching carefully. There are two separate issues here:

1) The focus-indicating `text-shadow` in the default skin is not a conventional focus indicator. It’s more of a light visual feedback for users whose eyes are already trained on the cursor. I realize mouse users and some designers sometimes find focus outlines uncool and cluttery, but I think it can be done both clearly and tastefully using conventional (dotted blue?) outlines. If a client violently hates the outlines, and an author can’t win that fight, then the author can turn them off, but clear, usable focus outlines should be incentivized in the default.

2) The visual leap across the player is unexpected, especially given the above.  Users expect tabbing to work consistently in one direction across a component. Obviously we don’t want to go setting tabindex on the elements. So one solution would be adding a wrapper for everything thats supposed to float right, float that wrapper right, and reverse the markup order of its children. Meh. Another slicker solution could be to switch the `float:right` elements to `display: inline-block`, and use `text-align:right`, and again reverse the markup order for those elements.

Thoughts?

**Comment by gdkraus** [comment]
This appears to be very similar to how YouTube handles it. YouTube appears to
1. assume a user is a mouse user
2. have a listener for key presses anywhere in the player, and on a key press add the class 'ytp-keyboard-focus' to the player, which has the corresponding :focus CSS to show the onscreen focus
3. have another listener for any mouse clicks in the player, and on a mouse click it removes the class 'ytp-keyboard-focus'

**Comment by mmcc** [comment]
Ok so just an update on this one, tab order has been immensely improved with the default theme for version 5. Instead of `float: right` we've moved to [flexbox](https://developer.mozilla.org/en-US/docs/Web/CSS/flex) for modern browsers and table as a fallback. Thanks to this combo the ordering stays intact, so we can now list the elements to be created in the correct order in the control bar and tab ordering stays intact.

That being said, I do think we should still improve the focus outlines, since even with the order being predictable it can still be difficult to see the focus shadow that's there right now. I like the YT-style approach @gdkraus (and @baloneysandwiches) mentioned. I'll probably submit my PR for the default theme without this work involved since that branch is already getting pretty massive, then we can handle this in a separate one.

**Comment by gkatsev** [comment]
Closing this. The tab ordering has been fixed in videojs 5 and we have a separate issue about the outlines (#3200).
